dataapi/float.go

```diff
@@ -0,0 +1,55 @@
+package dataapi
+
+import "fmt"
+
+// Float is a DataItem that wraps a float primitive
+type Float struct {
+	*BaseDataItem
+	value  float64
+	locked bool
+}
+
+// NewFloat returns a new Float dataItem
+func NewFloat(i float64) *Float {
+	return &Float{
+		BaseDataItem: NewBaseDataItem(),
+		value:        i,
+		locked:       false,
+	}
+}
+
+// String returns the string representation of the Float dataItem
+func (f *Float) String() string {
+	return fmt.Sprintf("%v", f.value)
+}
+
+// SetFloatWithMask - makes the Float dataItem implement 2 way binding
+// pass a newInt to set the dataItem, and set mask to ignore
+// the callback with the masked ID.  Pass a mask of 0 to trigger
+// all callbacks.
+func (f *Float) SetFloatWithMask(newfloat float64, mask int) {
+	if f.locked {
+		f.value = newfloat
+		return
+	}
```

Comment:
yes, same race condition as mentioned above.

Cant escape the problem of circular dependencies unfortunately.  Simple mutex locking was a good solution initially, but I think something more robust is needed, otherwise there will be race conditions all over the place.

I think it might need either a queue/channel or a map of both dataItems and widgets that are queued for propagation.  What do you think ?

The 7GUIs farenheit/celcius example highlights this issue as well, where updating the widget updates the opposing dataitem, which triggers widget updates.  An elegant way around this is to support computed properties, so the dataItem stores 1 temperature  (Kelvin), and then each widget uses a computed property / function to update the widget.

Reply:
Can you help me understand the limitation with mutex?

I implemented a version using mutex, listeners, and ensuring that every listener notification is run in a new goroutine.

7GUIs is a really good way to prove such an API. I like the idea of a single data store of kelvin which is listened to by convertors.